uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointsAliasMockMvcTests.java

```diff
@@ -327,229 +373,351 @@ protected UpdateBase(final boolean aliasFeatureEnabled) {
+                }
 
-                // check if the returned IdP has a redacted relying party secret
-                assertRelyingPartySecretIsRedacted(updatedIdp);
-            }
+                @Test
+                void shouldReject_IdpWithOriginKeyAlreadyPresentInOtherZone_CustomToUaaZone() throws Exception {
+                    shouldReject_IdpWithOriginKeyAlreadyPresentInOtherZone(customZone, IdentityZone.getUaa());
+                }
 
-            @Test
-            void shouldReject_OnlyAliasIdSet_UaaZone() throws Exception {
-                shouldReject_OnlyAliasIdSet(IdentityZone.getUaa());
-            }
+                private void shouldReject_IdpWithOriginKeyAlreadyPresentInOtherZone(final IdentityZone zone1, final IdentityZone zone2) throws Exception {
+                    // create IdP with origin key in zone 2
+                    final IdentityProvider<?> existingIdpInZone2 = buildOidcIdpWithAliasProperties(zone2.getId(), null, null);
+                    createIdp(zone2, existingIdpInZone2);
+
+                    // create IdP with same origin key in zone 1
+                    final IdentityProvider<?> idp = buildIdpWithAliasProperties(
+                            zone1.getId(),
+                            null,
+                            null,
+                            existingIdpInZone2.getOriginKey(), // same origin key
+                            OIDC10
+                    );
+                    final IdentityProvider<?> providerInZone1 = createIdp(zone1, idp);
+
+                    // update the alias ZID to zone 2, where an IdP with this origin already exists -> should fail
+                    providerInZone1.setAliasZid(zone2.getId());
+                    shouldRejectUpdate(zone1, providerInZone1, HttpStatus.CONFLICT);
+                }
 
-            @Test
-            void shouldReject_OnlyAliasIdSet_CustomZone() throws Exception {
-                shouldReject_OnlyAliasIdSet(customZone);
-            }
+                @Test
+                void shouldReject_AliasZidSetToSameZone_UaaZone() throws Exception {
+                    shouldReject_AliasZidSetToSameZone(IdentityZone.getUaa());
+                }
 
-            private void shouldReject_OnlyAliasIdSet(final IdentityZone zone) throws Exception {
-                final IdentityProvider<?> idp = buildOidcIdpWithAliasProperties(zone.getId(), null, null);
-                final IdentityProvider<?> createdProvider = createIdp(zone, idp);
-                assertThat(createdProvider.getAliasZid()).isBlank();
-                createdProvider.setAliasId(UUID.randomUUID().toString());
-                shouldRejectUpdate(zone, createdProvider, HttpStatus.UNPROCESSABLE_ENTITY);
-            }
+                @Test
+                void shouldReject_AliasZidSetToSameZone_CustomZone() throws Exception {
+                    shouldReject_AliasZidSetToSameZone(customZone);
+                }
```

Comment:
Looks good. Leaving comments to track my progress as this is challenging to read through.